Improve "sending a message to a service" tutorial (#8120)
ballerina-platform/ballerina-dev-website: swan-lake/integration-tutorials/sending-a-message-to-a-service.md

    ![Open diagram view](/learn/images/integration-tutorials/sending-a-message-to-a-service/open_diagram_view.gif)

> 3. Define a record type named `Doctor` corresponding to the payload from the backend service.

Reviewer: I don't think this is what @sameerajayasoma meant in Suggestion 2. We can describe how using records help with data validation, access, etc.

>     The payload from the backend service will be an array of JSON objects, where each JSON object will be similar to the following and can be modeled as `Doctor` record type.

Reviewer: I don't think this is what @sameerajayasoma meant in Suggestion 2. We can describe how using records help with data validation, access, etc.

Author: What I have understood here is that we better show _the reason for defining a record before other steps_ - Payload is a list of `Doctor`

Reviewer: IMO, the existing version already conveyed what we have now. 

> Generate a record type corresponding to the payload from the backend service using the ["Paste JSON as record"](https://wso2.com/ballerina/vscode/docs/references/convert-json-to-records/#via-the-command-palette) VS Code command by providing a sample of the expected JSON payload.
> 
> The payload from the backend service will be an array of JSON objects, where each JSON object will be similar to the following.

It will be good to include why. We can say that we can work with the JSON payload directly, but using records have these benefits...

Author: Updated. Please check

Author: ![image](https://github.com/ballerina-platform/ballerina-dev-website/assets/25485997/754364f6-c916-4201-81d7-6d89d4b958d2)

        "fee": 7000
    }
    ```
   
    ![Define a record](/learn/images/integration-tutorials/sending-a-message-to-a-service/define_record.gif)

    > **Note:**

>     > While it's possible to work with JSON payloads directly, utilizing record types in your code offers several advantages. Records provide enhanced type safety, validation capabilities, and automated code generation, and are maintainable when handling structured data.

Reviewer: ```suggestion
    > While it is possible to work with the JSON payload directly, utilizing record types in your code offers several advantages including enhanced type safety, data validation, and better tooling experience (e.g., completion).
```

4. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resource that accepts user requests, retrieves relevant details from the backend service, and responds to the request.

    - Open [Ballerina HTTP API Designer](https://wso2.com/ballerina/vscode/docs/design-the-services/http-api-designer) in VS Code

>    

Reviewer: "Define an HTTP resource that allow ..."

We are still not mentioning the return types?

        "fee": 7000
    }

>     ```

Reviewer: ````suggestion
    ```

````
Please make sure new lines are added appropriately.

>    ![Define a record](/learn/images/integration-tutorials/sending-a-message-to-a-service/define_record.gif)

Reviewer: Broken.

Author: Missed GIFs when committing

        "fee": 7000
    }
    ```
   
    ![Define a record](/learn/images/integration-tutorials/sending-a-message-to-a-service/define_record.gif)

    > **Note:**
    > While it is possible to work with the JSON payload directly, using record types offers several advantages including enhanced type safety, data validation, and better tooling experience (e.g., completion).

4. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resource that accepts user requests, retrieves relevant details from the backend service, and responds to the request.

    - Open the [Ballerina HTTP API Designer](https://wso2.com/ballerina/vscode/docs/design-the-services/http-api-designer) in VS Code
   

>     - Use `/healthcare` as the service path (or the context) of the service, which is attached to the listener listening on port `port`.

Reviewer: Needs to be updated, no longer have `port`.

Author: Oh Missed it. Thanks!

        "fee": 7000
    }
    ```
   
    ![Define a record](/learn/images/integration-tutorials/sending-a-message-to-a-service/define_record.gif)

    > **Note:**

>     > While it is possible to work with the JSON payload directly, utilizing record types in your code offers several advantages including enhanced type safety, data validation, and better tooling experience (e.g., completion).

Reviewer: ```suggestion
    > While it is possible to work with the JSON payload directly, using record types offers several advantages including enhanced type safety, data validation, and better tooling experience (e.g., completion).
```

        }
        ```

5. Define a [configurable variable](https://ballerina.io/learn/by-example/#configurability) for the URL of the backend service and an [`http:Client`](https://ballerina.io/learn/by-example/#http-client) object to send requests to the backend service.

>     ![Define a configurable and a client](/learn/images/integration-tutorials/sending-a-message-to-a-service/define_a_configurable_and_a_client.gif)

Reviewer: Broken.

configurable -> configurable variable

    The generated code will be as follows.

    ```ballerina
    configurable string healthcareBackend = "http://localhost:9090/healthcare";

>     final http:Client queryDoctorEP = check new (url = healthcareBackend);

Reviewer: Continuing the discussion re: https://github.com/ballerina-platform/ballerina-dev-website/pull/8057/files#r1374470440 here, I still think we should do it. Isn't it just a matter of selecting all (of the named arg), removing it, and just adding the variable reference?

Author: Found a way to do it properly. Please check

    - The `log:printInfo` statement [logs](https://ballerina.io/learn/by-example/#log) information about the request.

    - The call to the backend is done using a remote method call expression (using `->`), which distinguishes network calls from normal method calls. [Client data binding](https://ballerina.io/learn/by-example/http-client-data-binding/) is used to directly try and bind the JSON response on success to the expected array of records.

>     - Use the `is` check to decide the response based on the response to the client call. If the client call was successful and the respond payload was an array of `Doctor` records (as expected), then directly return the array from the resource. 

Reviewer: ```suggestion
    - Use the `is` check to decide the response based on the response to the backend call. If the backend call was successful and the response payload was an array of `Doctor` records (as expected), then directly return the array from the resource. 
```

>     - If the request fails, send an `http:NotFound` response if the client call failed with a `4xx` status code or send an `http:InternalServerError` response for other failures.

Reviewer: ```suggestion
    - If the backend call fails, send an `http:NotFound` response if the client call failed with a `4xx` status code or send an `http:InternalServerError` response for other failures.
```

>         Doctor[]|http:ClientError resp = queryDoctorEP->/[category];

Reviewer: Given the comments saying the code is kinda repeated, for this iteration, let's try removing the snippets from these descriptions and see.

    decimal fee;
|};

configurable string healthcareBackend = "http://localhost:9090/healthcare";

final http:Client queryDoctorEP = check new (healthcareBackend);

service /healthcare on new http:Listener(8290) {
    resource function get doctors/[string category]() 
            returns Doctor[]|http:NotFound|http:InternalServerError {

>         log:printInfo("Retrieving information", specialization = category);

Reviewer: Can you also update 

"Define a variable to keep the expected payload from the backend service" under testing to say something like

"Assign the mock payload from the backend service to a variable..."

"to keep" isn't ideal IMO.

>     final http:Client cl = check new (string `http://localhost:${port}/healthcare/doctors`);

Reviewer: Needs to be updated. No `port` now.

    ```

> 3. Assign the mock payload from the backend service to a variable. This variable will be used to mock the payload from the backend and to verify the received payload

Reviewer: ```suggestion
3. Define a variable with mock payload from the backend service. This variable will be used to mock the payload from the backend and to verify the received payload
```

>     final http:Client cl = check new (string `http://localhost:${port}/healthcare/doctors`);

Reviewer: "Step 3: Build and run the service" terminal has clear?

Author: Yeah. it automatically run a `clear`

>     final http:Client cl = check new (string `http://localhost:${port}/healthcare/doctors`);

Reviewer: Formatting seems to be broken with `bal run hospitalservice.jar`?

    ```

> 3. Assign the mock payload from the backend service to a variable.

Reviewer: We can still include the second sentence.
